hw3/hw3.playground/Contents.swift

```diff
@@ -0,0 +1,464 @@
+    var extraTires: Int = 5
+    
+    var steeringWheelSize: SteeringWheelSize = .L
+    
+    var wheelCount: Int = 4
+    
+    var steeringWheelColor: UIColor
+    
+    var enginePower: Int = 75
+    
+    var engineState: EngineState = .wellDone
+    
+    var country: Location
+    
+    var manufacture: Manufacture
+    
+    var vin: String
+    
+    var state: CarState = .prepared
+    
+    var factory: CarFactory
+    
+    init(country: Location, manufacture: Manufacture, vin: String, factory: CarFactory, color: UIColor) {
+        self.country = country
+        self.manufacture = manufacture
+        self.vin = vin
+        self.factory = factory
+        steeringWheelColor = color
+    }
+    
+    func offRoadMode(){//Witness Table Dispatch
+        print("Режим бездорожья активирован")
+    }
+    var nitro:Int  = 2
+}
+
+
+public protocol Dillerable {
+    
+    var sedanCarParking:[SedanCar]{get set}
+    
+    var sportCarParking:[SportCar]{get set}
+    
+    var vanCarParking:[VanCar] {get set}
+    
+    var crossOverCarParking:[CrossOverCar]{get set}
+    
+    mutating func addCar(car:SedanCar)
+    
+    mutating func addCar(car:SportCar)
+    
+    mutating func addCar(car:VanCar)
+    
+    mutating func addCar(car:CrossOverCar)
+    
+    mutating  func sellByVin(vin:String)
+}
+
+
+public extension Dillerable {
+    
+    mutating func sellByVin(vin:String){//Static(Direct) Dispatch
+        if var car = sportCarParking.first(where: {$0.vin == vin}) {
+            car.sell()
+            sportCarParking.removeAll(where: {$0.vin == vin})
+        }
+        if var car = crossOverCarParking.first(where: {$0.vin == vin}) {
+            car.sell()
+            crossOverCarParking.removeAll(where: {$0.vin == vin})
+        }
+        if var car = vanCarParking.first(where: {$0.vin == vin}) {
+            car.sell()
+            vanCarParking.removeAll(where: {$0.vin == vin})
+        }
+        if var car = sportCarParking.first(where: {$0.vin == vin}) {
+            car.sell()
+            sportCarParking.removeAll(where: {$0.vin == vin})
+        }
+    }
+    
+    mutating func addCar(car:SedanCar){//Static(Direct) Dispatch
+        sedanCarParking.append(car)
+    }
+    
+    mutating func addCar(car:SportCar){//Static(Direct) Dispatch
+        sportCarParking.append(car)
+    }
+    
+    mutating func addCar(car: VanCar){//Static(Direct) Dispatch
+        vanCarParking.append(car)
+    }
+    
+    mutating func addCar(car:CrossOverCar){//Static(Direct) Dispatch
+        crossOverCarParking.append(car)
+    }
+    
+}
+
+public class Diller:Dillerable{
+    
+    public var sedanCarParking: [SedanCar] = []
+    
+    public var sportCarParking: [SportCar] = []
+    
+    public var vanCarParking: [VanCar] = []
+    
+    public var crossOverCarParking: [CrossOverCar] = []
+    
+    init(){
+    }
+}
+
+public class CarFactory{
+    private var name:String
+    var diller: Diller?
```

Comment:
Не совсем, у фабрики должен быть делегат, а делегатом может быть любой объект ссылочного типа.
(Диллер, экспортер, импортер и тд)
Поэтому фабрика должна взаимодействовать со своим делегатом на уровне протокола.
Нужно создать протокол FactoryDelegate и уже Диллера подписать на этот протокол.
В протоколе уже указать например методы по передаче делегату изготовленных автомобилей.
```suggestion
    weak var delegate: FactoryDelegate?
```